### PR TITLE
GH-39231: [C++][Compute] Add binary_slice kernel for fixed size binary

### LINKED ISSUE #39231: [C++][Compute] binary_slice function should support fixed-size binary arrays
### Describe the enhancement requested

It would be nice to have a quick way to truncate or enlarge a fixed-size binary array, for example go from `fixed_size_binary<16>` to `fixed_size_binary<7>` or the converse.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by js8544**
### Rationale for this change
Add binary_slice kernel for fixed size binary

### What changes are included in this PR?
Add binary_slice kernel for fixed size binary

### Are these changes tested?
Yes

### Are there any user-facing changes?
No

* Closes: #39231

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_string_ascii.cc [hunk 1]
```diff
@@ -2568,6 +2571,32 @@ struct SliceBytesTransform : StringSliceTransformBase {
...
+    start = std::min(start, input_width);
+    stop = std::min(stop, input_width);
+
+    if ((start == stop) || ((start >= stop) && (step > 0)) ||
+        ((start <= stop) && (step < 0))) {
+      return 0;
+    }
+    return static_cast<int32_t>(std::max(
+        static_cast<int64_t>(0), (stop - start + (step - (step > 0 ? 1 : -1))) / step));
+  }
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_ascii.cc:2599]: Hmm, can you rephrase the code such that `step > 0` and `step < 0` have separate code paths? It would make things clearer.

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 2]
```diff
@@ -712,6 +712,95 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
              "[6, null, 6]");
 }
 
+TEST_F(TestFixedSizeBinaryKernels, SliceBytesBasic) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:715]: Nit
```suggestion
TEST_F(TestFixedSizeBinaryKernels, BinarySliceBasic) {
```

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 3]
```diff
@@ -712,6 +712,95 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
...
+  SliceOptions options{2, 4};
+  CheckUnary("binary_slice", R"(["abcabc", "defdef"])", fixed_size_binary(2),
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:717]: Can we use less repeated values to make sure the substring selection is right?
```suggestion
  CheckUnary("binary_slice", R"(["abcdef", "ghijkl"])", fixed_size_binary(2),
```

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc:717]: Also, can you add some null values at some point?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for tackling this @js8544 ! Here are some comments.

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc (hunk 2 above)
Comment (js8544) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc]: Done

File: cpp/src/arrow/compute/kernels/scalar_string_ascii.cc (hunk 1 above)
Comment (js8544) [reply cpp/src/arrow/compute/kernels/scalar_string_ascii.cc]: Done

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc (hunk 3 above)
Comment (js8544) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc]: Done.

File: cpp/src/arrow/compute/kernels/scalar_string_ascii.cc [hunk 4]
```diff
@@ -95,7 +95,7 @@ struct FixedSizeBinaryTransformExecBase {
                           ctx->Allocate(output_width * input_nstrings));
     uint8_t* output_str = values_buffer->mutable_data();
 
-    const uint8_t* input_data = input.GetValues<uint8_t>(1);
+    const uint8_t* input_data = input.GetValues<uint8_t>(1, input.offset * input_width);
```
Comment (js8544) [inline cpp/src/arrow/compute/kernels/scalar_string_ascii.cc:98]: The current `FixedSizeBinaryTransformExecBase` deals with offsets wrongly. I fixed the bug and added a test for `binary_replace_slice` as well.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/scalar_string_ascii.cc:98]: Nice catch, thank you!

**Comment by pitrou** [comment]
I think this PR produces incorrect results sometimes:
```python
>>> pc.binary_slice(pa.array([b"ab\xf2de"], pa.binary(5)), -5, -6, -3)
Traceback (most recent call last):
  Cell In[16], line 1
    pc.binary_slice(pa.array([b"ab\xf2de"], pa.binary(5)), -5, -6, -3)
  File ~/arrow/dev/python/pyarrow/compute.py:263 in wrapper
    return func.call(args, options, memory_pool)
  File pyarrow/_compute.pyx:385 in pyarrow._compute.Function.call
    result = GetResultValue(
  File pyarrow/error.pxi:154 in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File pyarrow/error.pxi:91 in pyarrow.lib.check_status
    raise convert_status(status)
ArrowInvalid: Invalid UTF8 sequence in input
```

This is when adding this validation test in Python:
```diff
diff --git a/python/pyarrow/tests/test_compute.py b/python/pyarrow/tests/test_compute.py
index 7c5a134d33..d1eb605c71 100644
--- a/python/pyarrow/tests/test_compute.py
+++ b/python/pyarrow/tests/test_compute.py
@@ -561,7 +561,8 @@ def test_slice_compatibility():
 
 
 def test_binary_slice_compatibility():
-    arr = pa.array([b"", b"a", b"a\xff", b"ab\x00", b"abc\xfb", b"ab\xf2de"])
+    data = [b"", b"a", b"a\xff", b"ab\x00", b"abc\xfb", b"ab\xf2de"]
+    arr = pa.array(data)
     for start, stop, step in itertools.product(range(-6, 6),
                                                range(-6, 6),
                                                range(-3, 4)):
@@ -574,6 +575,13 @@ def test_binary_slice_compatibility():
         assert expected.equals(result)
         # Positional options
         assert pc.binary_slice(arr, start, stop, step) == result
+        # Fixed size binary input / output
+        for item in data:
+            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
+            expected = item[start:stop:step]
+            actual = pc.binary_slice(fsb_scalar, start, stop, step)
+            assert actual.type == pa.binary(len(expected))
+            assert actual.as_py() == expected
 
 
 def test_split_pattern():
```

**Comment by js8544** [comment]
There was a bug in my implementation. I refactored and reused the range computing logic in the kernel exec function. And I also added tests like the one you gave above, in both C++ and Python.

File: python/pyarrow/tests/test_compute.py [hunk 5]
```diff
@@ -574,6 +575,16 @@ def test_binary_slice_compatibility():
         assert expected.equals(result)
         # Positional options
         assert pc.binary_slice(arr, start, stop, step) == result
+        # Fixed size binary input / output
+        for item in data:
+            print(item)
+            print(start, stop, step)
+            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
+            expected = item[start:stop:step]
+            print(expected)
```
Comment (pitrou) [inline python/pyarrow/tests/test_compute.py:584]: Remove the print statements here?
```suggestion
            fsb_scalar = pa.scalar(item, type=pa.binary(len(item)))
            expected = item[start:stop:step]
```

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 6]
```diff
@@ -712,11 +712,139 @@ TEST_F(TestFixedSizeBinaryKernels, BinaryLength) {
...
+          if (step == 0) {
+            continue;
+          }
+          SliceOptions options{start, stop, step};
+          auto expected =
+              CallFunction("binary_slice", {varlen_input}, &options).ValueOrDie();
+          auto actual =
+              CallFunction("binary_slice", {fixed_input}, &options).ValueOrDie();
+          actual = Cast(actual, binary()).ValueOrDie();
+          AssertDatumsEqual(expected, actual);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:836]: We should validate the output:
```suggestion
          ASSERT_OK(actual.make_array()->ValidateFull());
          AssertDatumsEqual(expected, actual);
```

**Comment by js8544** [comment]
Updated. Thanks for pointing out!

**Review by pitrou — APPROVED** [review]
